rust/connlib/tunnel/src/client.rs

```diff
@@ -648,10 +649,28 @@ impl ClientState {
         &mut self,
         query: DnsQuery<'static>,
         response: Result<
-            Result<hickory_resolver::lookup::Lookup, hickory_resolver::error::ResolveError>,
-            futures_bounded::Timeout,
+            Result<
+                Result<hickory_resolver::lookup::Lookup, hickory_resolver::error::ResolveError>,
+                futures_bounded::Timeout,
+            >,
+            DnsQueryError,
         >,
     ) {
+        let response = match response {
+            Ok(response) => response,
+            Err(e) => {
+                tracing::warn!(name = %query.name, server = %query.query.destination(), "Failed to send DNS query: {e}");
+                let response = ip_packet::make::dns_err_response(
+                    query.query,
+                    hickory_proto::op::ResponseCode::ServFail,
+                )
+                .map(|p| p.into_immutable());
+
+                self.buffered_packets.extend(response);
+                return;
+            }
+        };
```

Comment:
And then this block converts it back to the original `Result <HickoryLookup, HickoryResolveError>, Timeout>` type by stripping out the `DnsQueryError` and using it to send (buffer) an error response packet, right? I assume all the other branches except maybe timeout ultimately buffer a response packet, too.

Reply:
Basically, what these `match` blocks do is deal with one layer of error at a time, kind of "early-exit" style. Technically, we could also collapse those. I'll see what that looks like.